### PR TITLE
[TR] PIM-5375: Test migration from 1.4 to 1.5

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                 | A
| ----------------- | ---
| Specs             | ok
| Behats            | ok
| Blue CI           |
| Changelog updated | ok
| Review and 2 GTM  |
| Micro Demo (PO)   | no need
| Migration script  | ok
| Tech Doc          |
Link to PR in std version : https://github.com/akeneo/pim-community-standard/pull/259

In order to not have anymore to move and adapt migration scripts from dev to std for each releases, we decided to create a pre-update-cmd script for composer which will copy `upgrades/` folder from dev to std.
So we have to repatriate all the `upgrades/` folder from std to dev. Then, do the composer script (actually in the class `InstallerBundle\ComposerScripts`)

:warning:  Pls, don't review files which are in `upgrades/1.x-1.x/` as they are old migration scripts, and I'll not fix all those scripts. Else I'll have to re-test all migrations :trollface: 

- [x] Move upgrades/
- [x] Add script
- [x] Copy UPGRADE.md with the script
- [x] Clean UPGRADE-1.5.md
- [x] Move three classes which are in TODO in the UPGRADE.md sed part.
- [x] Test ORM
- [x] Test ODM
- [x] Move ABOUT-1.5.md in tech doc

You can retrieve all steps in commits to ease the review.

### CODE REVIEW

File: UPGRADE-1.5.md [hunk 1]
```diff
@@ -4,487 +4,64 @@
 
 > Please perform a backup of your codebase if you don't use any VCS.
 
-## PHP7 and HHVM
+## Update dependencies and configuration
 
-We continue to do several compatibility tests for PHP7 and HHVM.
```
Comment (solivier) [inline UPGRADE-1.5.md]: Why did you remove that ?

File: src/Pim/Bundle/CatalogBundle/Resources/config/entities.yml [hunk 2]
```diff
@@ -26,4 +26,4 @@ parameters:
     pim_catalog.entity.metric.class:                      Pim\Component\Catalog\Model\Metric
     pim_catalog.entity.product_price.class:               Pim\Component\Catalog\Model\ProductPrice
     pim_catalog.entity.product_template.class:            Pim\Bundle\CatalogBundle\Entity\ProductTemplate
-    pim_catalog.entity.available_attributes.class:        Pim\Component\Catalog\Model\AvailableAttributes
+    pim_catalog.entity.available_attributes.class:        Pim\Component\Enrich\Model\AvailableAttributes
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/entities.yml]: is the namespace has changed I think you should change the service parameter name too aka : pim_enrich.entity.available_attributes.class

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+namespace Pim\Bundle\InstallerBundle;
+
+use Composer\Script\Event;
+use Pim\Bundle\CatalogBundle\Version;
+
+/**
+ * Contains all PIM composer scripts we run.
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (solivier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: 2016 :trollface:

File: upgrades/1.1-1.2/mongodb/migrate_media.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
+<?php
+
+$productCollName = 'pim_catalog_product';
+$mediaCollName = 'pim_catalog_media';
+
+if (count($argv) !== 3) {
```
Comment (solivier) [inline upgrades/1.1-1.2/mongodb/migrate_media.php]: yoda style

File: upgrades/1.1-1.2/mongodb/migrate_media.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        ['values._id' => $valueId],
+        [
+            '$set' => [
+                'values.$.media' => $media
+            ]
+        ],
+        ['w' => true]
+    );
+
+    if ($result['ok'] != 1) {
```
Comment (solivier) [inline upgrades/1.1-1.2/mongodb/migrate_media.php]: yoda style

File: upgrades/1.1-1.2/orm/migrate_versions.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+require_once __DIR__.'/../../../app/AppKernel.php';
+
+$input = new ArgvInput($argv);
+$env = $input->getParameterOption(['-e', '--env']);
+if (!$env) {
+    echo sprintf("Usage: %s --env=<environment>\nExample: %s --env=dev\n", $argv[0], $argv[0]);
+    exit(1);
+}
+
+$kernel = new AppKernel($env, $env === 'dev');
```
Comment (solivier) [inline upgrades/1.1-1.2/orm/migrate_versions.php]: dev ? is it normal ?

File: UPGRADE-1.5.md (hunk 1 above)
Comment (willy-ahva) [inline UPGRADE-1.5.md]: nidup did some documentation for the new 1.5 when we start to work on it. This documentation is not part of the migration processus but more technical explaination about what we did and why. I removed this doc and added it in pim-doc. Here is the PR https://github.com/akeneo/pim-docs/pull/219

Comment (solivier) [inline UPGRADE-1.5.md]: ah ok nice :+1:

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    public static function copyUpgradesDirFromDev(Event $event)
+    {
+        $event->getIO()->write('Copying upgrades folder from Akeneo PIM dependency to standard version.');
+        $ds = DIRECTORY_SEPARATOR;
+
+        $vendorDir   = $event->getComposer()->getConfig()->get('vendor-dir');
+        $upgradesDir = 'upgrades' . $ds;
+        $akeneoDir   = $vendorDir . $ds . 'akeneo';
+
+        if ('CE' === Version::EDITION) {
```
Comment (solivier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: as you import the 'Pim\Bundle\CatalogBundle\Version' it will be always the string 'CE' no ?

**Comment by solivier** [comment]
GTM

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php (hunk 7 above)
Comment (willy-ahva) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: Nope it can be the override in EE

Comment (solivier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: yes it can be overriden in EE but as you import 'Pim\Bundle\CatalogBundle\Version' instead of 'PimEnterprise\Bundle\CatalogBundle\Version' and as it's a static it will always be 'CE'

Comment (willy-ahva) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: yes you right I need to refactor this ty

File: UPGRADE-1.5.md [hunk 8]
```diff
@@ -4,487 +4,64 @@
...
 
-The "flat/csv" format should be only used for import/export and could reside in Connector component:
-Pim/Bundle/TransformBundle/Normalizer/Flat -> Pim/Component/Connector/Normalizer/Flat
- -> because should only be used for csv import/export (and currently versioning for legacy reasons)
-
-The "internal_api" format could be renamed in Enrich bundle:
- -> Pim/Bundle/EnrichBundle/Normalizer -> Pim/Bundle/EnrichBundle/Normalizer/InternalRest
-
-Other bundles register normalizers/denormalizers for these formats and could be re-organized:
+Don't forget to perform a backup of your database before !
```
Comment (jjanvier) [inline UPGRADE-1.5.md]: you can remove this line
we already warned the user at the very beginning of the user

File: UPGRADE-1.5.md [hunk 9]
```diff
@@ -4,487 +4,64 @@
...
- - The parameters and services 'pim_catalog.mongodb.normalizer.*' have been renamed to 'pim_catalog.mongodb.normalizer.normalized_data.*'
- - The "mongodb_document" format has been moved in Catalog Bundle (not in component because rely on storage classes): Pim/Bundle/TransformBundle/Normalizer/MongoDB -> Pim/Bundle/CatalogBundle/Normalizer/MongoDB/Document
- - The parameters and services 'pim_serializer.normalizer.mongodb.*' have been renamed to 'pim_catalog.mongodb.normalizer.document.*'
-
-TODO:
+This step will also copy the upgrades folder from `vendor/akeneo/pim-community-dev/` to your Pim project root to allow you to migrate.
 
-The "structured/json/standard" format could be moved to Catalog component:
- - Pim/Bundle/TransformBundle/Normalizer/Structured -> Pim/Component/Catalog/Normalizer/Structured
+Then you can migrate your database using :
```
Comment (jjanvier) [inline UPGRADE-1.5.md]: by using:

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        if (!file_exists($copyFromPath)) {
+            $event->getIO()->writeError(sprintf(
+                'Directory "%s" not found. Update will continue but no Akeneo PIM migration can be done.',
+                $copyFromPath
+            ));
+
+            return;
+        }
+
+        if (file_exists($copyToPath) && $rm) {
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: I think we should use Symfony components here

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        $event->getIO()->write('Done.');
+    }
+
+    /**
+     * @param Event  $event        Composer script event
+     * @param string $copyFromPath Copy files from this folder path
+     * @param string $copyToPath   Copy files to this folder path
+     * @param bool   $rm           Remove the destination folder if it already exists
+     */
+    protected static function copyFiles(Event $event, $copyFromPath, $copyToPath, $rm = true)
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: the $rm argument is uselss
you don't use it anywhere in the code as it's always set to true

you should remove it imo

**Comment by jjanvier** [comment]
@willy-ahva one part of the "Catalog Bundle & Component [WIP]" section is mandatory in the UPGRADE.md

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php (hunk 11 above)
Comment (willy-ahva) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: used in EE

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ComposerScripts
+{
+    /**
+     * @param Event $event
+     */
+    public static function copyMigrationFolderFromDev(Event $event)
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: I would name it `easeAkeneoPimUpgrade` imo, or something like that
don't really know but it does not only copy Doctrine migrations

Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: shouldn't you here also copy all the UPGRADE*.md files? just wondering

Comment (willy-ahva) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: Yes I could. It could avoid to copy paste the upGRADE.md too. And then rename the method. Sounds ok for me

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php (hunk 10 above)
Comment (willy-ahva) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: Doing it

**Comment by willy-ahva** [comment]
@jjanvier done

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+     * Copy the upgrades/ folder from PIM dependency to the standard edition
+     *
+     * @param Event $event
+     */
+    protected static function copyUpgradesFolder(Event $event)
+    {
+        $cePath        = static::getDependencyMigrationPath($event, 'pim-community-dev');
+        $migrationPath = static::getMigrationFolderPath($event);
+
+        self::copyFiles($event, $cePath, $migrationPath);
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: don't mix self and static, choose one

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    protected static function copyUpgradeFile(Event $event)
+    {
+        $ds = DIRECTORY_SEPARATOR;
+
+        $cePath        = static::getDependencyMigrationPath($event, 'pim-community-dev');
+        $ceDepPath     = $cePath . '..' . $ds;
+        $ceRoot        = $event->getComposer()->getConfig()->get('vendor-dir') . $ds . '..' . $ds;
+
+        $files = scandir($ceDepPath);
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: use the Symfony Finder component to do such things please

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+                $copyFromPath
+            ));
+
+            return;
+        }
+
+        if (is_dir($copyFromPath)) {
+            static::sanitizePath($destinationPath);
+
+            $files = scandir($copyFromPath);
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: use the Symfony Finder component (maybe it can give all files recursively, to test)

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+            $fs->copy($copyFromPath, $destinationPath, true);
+        }
+    }
+
+    /**
+     * Sanitize the given path. Create the folder if it doesn't exist.
+     *
+     * @param string $path
+     */
+    protected static function sanitizePath($path)
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: it does not sanitize a path, it creates a dir

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        $finder = new Finder();
+        $ds     = DIRECTORY_SEPARATOR;
+
+        $cePath    = static::getDependencyMigrationPath($event, 'pim-community-dev');
+        $ceDepPath = $cePath . '..' . $ds;
+        $ceRoot    = $event->getComposer()->getConfig()->get('vendor-dir') . $ds . '..' . $ds;
+
+        $finder->files()->name('UPGRADE-*.md');
+        foreach ($finder->in($ceDepPath) as $file) {
+            var_dump($file);
```
Comment (skeleton) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: var_dump ?

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+            return;
+        }
+
+        if (is_dir($copyFromPath)) {
+            $finder = new Finder();
+            static::sanitizePath($destinationPath);
+
+            foreach ($finder->in($copyFromPath) as $file) {
+                var_dump($file);
```
Comment (skeleton) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: same

File: src/Pim/Bundle/InstallerBundle/ComposerScripts.php [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        $ds     = DIRECTORY_SEPARATOR;
+
+        $cePath    = static::getDependencyMigrationPath($event, 'pim-community-dev');
+        $ceDepPath = $cePath . '..' . $ds;
+        $ceRoot    = $event->getComposer()->getConfig()->get('vendor-dir') . $ds . '..' . $ds;
+
+        $finder
+            ->files()
+            ->name('UPGRADE-*.md')
+            ->notPath('vendor')
```
Comment (jjanvier) [inline src/Pim/Bundle/InstallerBundle/ComposerScripts.php]: depth('== 0')
